engines/lab/map.cpp

```diff
@@ -0,0 +1,577 @@
+		drawY = y + (_imgRoom->_height - _imgMapX[_direction]->_height) / 2;
+
+		break;
+
+	case BRIDGEROOM:
+		_imgBridge->drawImage(x, y);
+
+		drawX = x + (_imgBridge->_width - _imgMapX[_direction]->_width) / 2;
+		drawY = y + (_imgBridge->_height - _imgMapX[_direction]->_height) / 2;
+
+		break;
+
+	case VCORRIDOR:
+		_imgVRoom->drawImage(x, y);
+
+		offset = (_imgVRoom->_width - _imgPath->_width) / 2;
+
+		if (NORTHDOOR & flags)
+			_imgPath->drawImage(x + offset, y - _imgPath->_height);
+
+		if (SOUTHDOOR & flags)
+			_imgPath->drawImage(x + offset, y + _imgVRoom->_height);
+
+		offset = (_imgRoom->_height - _imgPath->_height) / 2;
+
+		if (EASTDOOR & flags)
+			_imgPath->drawImage(x + _imgVRoom->_width, y + offset);
+
+		if (WESTDOOR & flags)
+			_imgPath->drawImage(x - _imgPath->_width, y + offset);
+
+		if (EASTBDOOR & flags)
+			_imgPath->drawImage(x + _imgVRoom->_width, y - offset - _imgPath->_height + _imgVRoom->_height);
+
+		if (WESTBDOOR & flags)
+			_imgPath->drawImage(x - _imgPath->_width, y - offset - _imgPath->_height + _imgVRoom->_height);
+
+		offset = (_imgVRoom->_height - _imgPath->_height) / 2;
+
+		if (EASTMDOOR & flags)
+			_imgPath->drawImage(x + _imgVRoom->_width, y - offset - _imgPath->_height + _imgVRoom->_height);
+
+		if (WESTMDOOR & flags)
+			_imgPath->drawImage(x - _imgPath->_width, y - offset - _imgPath->_height + _imgVRoom->_height);
+
+		drawX = x + (_imgVRoom->_width - _imgMapX[_direction]->_width) / 2;
+		drawY = y + (_imgVRoom->_height - _imgMapX[_direction]->_height) / 2;
+
+		break;
+
+	case HCORRIDOR:
+		_imgHRoom->drawImage(x, y);
+
+		offset = (_imgRoom->_width - _imgPath->_width) / 2;
+
+		if (NORTHDOOR & flags)
+			_imgPath->drawImage(x + offset, y - _imgPath->_height);
+
+		if (SOUTHDOOR & flags)
+			_imgPath->drawImage(x + offset, y + _imgRoom->_height);
+
+		if (NORTHRDOOR & flags)
+			_imgPath->drawImage(x - offset - _imgPath->_width + _imgHRoom->_width, y - _imgPath->_height);
+
+		if (SOUTHRDOOR & flags)
+			_imgPath->drawImage(x - offset - _imgPath->_width + _imgHRoom->_width, y + _imgRoom->_height);
+
+		offset = (_imgHRoom->_width - _imgPath->_width) / 2;
+
+		if (NORTHMDOOR & flags)
+			_imgPath->drawImage(x - offset - _imgPath->_width + _imgHRoom->_width, y - _imgPath->_height);
+
+		if (SOUTHMDOOR & flags)
+			_imgPath->drawImage(x - offset - _imgPath->_width + _imgHRoom->_width, y + _imgRoom->_height);
+
+		offset = (_imgRoom->_height - _imgPath->_height) / 2;
+
+		if (EASTDOOR & flags)
+			_imgPath->drawImage(x + _imgHRoom->_width, y + offset);
+
+		if (WESTDOOR & flags)
+			_imgPath->drawImage(x - _imgPath->_width, y + offset);
+
+		drawX = x + (_imgHRoom->_width - _imgMapX[_direction]->_width) / 2;
+		drawY = y + (_imgHRoom->_height - _imgMapX[_direction]->_height) / 2;
+
+		break;
+
+	default:
+		return;
+	}
+
+	if (drawMarkFl)
+		_imgMapX[_direction]->drawImage(drawX, drawY);
+}
+
+bool LabEngine::floorVisited(uint16 floorNum) {
+	for (uint16 i = 1; i <= _maxRooms; i++) {
```

Comment:
This looks strange: _maps[] is allocated with "_maps = new MapData[_maxRooms];" and here we access elements 1 through _maxRooms. It may be ok to ignore element 0, but element _maxRooms is outside of the allocated range.